### DISCUSSION TITLE
Why doesn't React on Rails hydrate at the `DOMContentLoaded` event?

### DISCUSSION THREAD

**Discussion by kotarella1110** [shakacode/react_on_rails#1665] [Q&A] [answered]
In React on Rails, hydration and rendering sometimes occur at the `load` event. This happens because of the current implementation of the `onPageReady` function.  

https://github.com/shakacode/react_on_rails/blob/4339d30fc91eb4d16f0d9ad9cb430b24f9bed5c8/node_package/src/clientStartup.ts#L310-L337

~~If `clientStartup` is executed when `readyState` is `loading`, hydration occurs at the `interactive` state ([`DOMContentLoaded`](https://developer.mozilla.org/docs/Web/API/Document/DOMContentLoaded_event) event). However, if `clientStartup` runs when `readyState` is already `interactive`, hydration is delayed until the `complete` state ([`load`](https://developer.mozilla.org/docs/Web/API/Window/load_event) event).~~ => **When `clientStartup` is executed, hydration always occurs when readyState is `complete`.**

This behavior causes hydration to wait until all sub-resources, such as third-party JavaScript or Google Fonts, are fully loaded, potentially delaying the user’s interaction readiness.  

Would it be possible to allow hydration at the `DOMContentLoaded` event when the `readyState` is already `interactive`? For example, the following update to `onPageReady` could address this:  

```diff
function onPageReady(callback: () => void) {
- if (document.readyState === "complete") {
+ if (document.readyState !== "loading") {
    callback();
  } else {
    document.addEventListener("readystatechange", function onReadyStateChange() { 
      onPageReady(callback); 
      document.removeEventListener("readystatechange", onReadyStateChange); 
    });
  }
}
```

This adjustment ensures hydration begins as soon as the `DOMContentLoaded` event fires, even if `clientStartup` is executed after the `interactive` state.  

Are there any reasons or side effects that might prevent this change from being implemented? If this approach is acceptable, I’d be happy to submit a PR for it.

**Comment by alexeyr** [comment]
As it happens, I was looking into this as well.

It used to be on `DOMContentLoaded`, but was switched to `load`: https://github.com/shakacode/react_on_rails/pull/1540, mentioning an issue with `DOMContentLoaded` was revealed in https://github.com/shakacode/react_on_rails/pull/1539.

And that mentions https://github.com/shakacode/react_on_rails/commit/7b301f9e87b4f27b9197207f8e210c38095c551b which changed it the other way around earlier...

@Judahmeek @justin808 Do you remember what exactly "the race condition revealed by https://github.com/shakacode/react_on_rails/pull/1539" was? And was the intention actually "after `DOMContentLoaded`" or "on `load`"?

> If clientStartup is executed when readyState is loading, hydration occurs at the interactive state ([DOMContentLoaded](https://developer.mozilla.org/docs/Web/API/Document/DOMContentLoaded_event) event).

I don't think that's correct; when the state changes to `interactive`, it just adds `onReadyStateChange` as a listener again, and the callback will only run when it's `complete`.

**Reply by kotarella1110** [reply]
> I don't think that's correct; when the state changes to interactive, it just adds onReadyStateChange as a listener again, and the callback will only get run when it's complete.

My understanding was incorrect. You’re absolutely right. It seems that hydration always happens at the `onload` event.  
I’m currently facing this issue and looking for a way to address it because the hydration process is too slow.
By the way, in our current product using React on Rails, it takes about 4 seconds to load all sub-resources after `DOMContentLoaded`. This means there is a significant delay before users can interact with the page.

**Reply by justin808** [reply]
Should we make this configurable? Or is there a way that we can choose sensible defaults?

**Reply by AbanoubGhadban** [reply]
@kotarella1110 If you're confident that the component you're hydrating is already registered before the `onload` event, have you tried manually invoking the `ReactOnRails.reactOnRailsPageLoaded` function? Additionally, we're planning to release an update very soon that will allow you to hydrate any component on the page as soon as its corresponding script is loaded and executed. This update eliminates the need to wait for either the `onload` or `DOMContentLoaded` events.

**Reply by Judahmeek** [reply]
Another option for @kotarella1110 to consider would be some sort of lazy loading, although I'm not sure how lazy it would have to be to make sure that the browser doesn't consider it a sub-resource.

**Reply by kotarella1110** [reply]
@AbanoubGhadban 

> If you're confident that the component you're hydrating is already registered before the onload event, have you tried manually invoking the `ReactOnRails.reactOnRailsPageLoaded` function?

I wasn’t aware of the `ReactOnRails.reactOnRailsPageLoaded` function! I’ll definitely give it a try.
However, If I manually invoke the `ReactOnRails.reactOnRailsPageLoaded` function before the onload event, wouldn’t the `onPageReady` function potentially re-execute it again at the onload timing? In other words, I’m concerned that a hydration process that should only run once might end up being executed twice.

> Additionally, we're planning to release an update very soon that will allow you to hydrate any component on the page as soon as its corresponding script is loaded and executed.

That sounds fantastic. I’m looking forward to it!

@Judahmeek 

> Another option for @kotarella1110 to consider would be some sort of lazy loading, although I'm not sure how lazy it would have to be to make sure that the browser doesn't consider it a sub-resource.

I’ve considered lazy loading for sub-resources, but I don’t think it’s realistic in our case since some scripts can be lazy-loaded while others cannot. Additionally, significantly delaying font loading could lead to FOUT, which we’d like to avoid.

**Reply by kotarella1110** [reply]
Additionally, I’d like to understand the reasoning behind changing the hydration timing from the `DOMContentLoaded` event to the `load` event. I understand that this was related to addressing a race condition, but could you share more details about the specific issues that arose?

**Reply by Judahmeek** [reply]
@kotarella1110 the problem is that the `DOMContentLoaded` event fires once the HTML document has been loaded, but before all sub-resources, such as the react components in the webpack bundles, have been loaded. Which meant that relying on the `DOMContentLoaded` event could cause ReactOnRails to try to initialize before all the react components had been registered, which would then cause a fatal error.

The `load` event only fires once all sub-resources have been loaded.

**Reply by kotarella1110** [reply]
Does my understanding align with the problem description that `renderInit` might execute before `ReactOnRails.register` is completed? 

If so, could this issue be resolved by ensuring `renderInit` waits for `ReactOnRails.register` to finish? Below is a sample implementation using a `Deferred` mechanism:

```ts
// node_package/src/Deferred.ts
export class Deferred<T> {
  promise: Promise<T>;
  resolve!: (value: T | PromiseLike<T>) => void;
  reject!: (reason?: any) => void;

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  }
}
```

```ts
// node_package/src/reactOnRailsRenderedDefer.ts
import { Deferred } from './Deferred';

export const reactOnRailsRenderedDefer = new Deferred<void>();
```

```ts
// node_package/src/ReactOnRails.ts
import { reactOnRailsRenderedDefer } from './reactOnRailsRenderedDefer';
// ...
ctx.ReactOnRails = {
  // ...
  register(components: { [id: string]: ReactComponentOrRenderFunction }): void {
    ComponentRegistry.register(components);
    reactOnRailsRenderedDefer.resolve();
  },
  // ...
};
```

```ts
// node_package/src/clientStartup.ts
import { reactOnRailsRenderedDefer } from './reactOnRailsRenderedDefer';
// ...
async function renderInit(): Promise<void> {
  await reactOnRailsRenderedDefer.promise;
  // Install listeners when running on the client (browser).
  // We must do this check for turbolinks AFTER the document is loaded because we load the
  // Webpack bundles first.
  if ((!turbolinksInstalled() || !turbolinksSupported()) && !turboInstalled()) {
    debugTurbolinks('NOT USING TURBOLINKS: calling reactOnRailsPageLoaded');
    reactOnRailsPageLoaded();
    return;
  }
  // ...
}
// ...
function onPageReady(callback: () => void) {
  if (document.readyState !== "loading") {
    callback();
  } else {
    document.addEventListener("DOMContentLoaded", callback);
  }
}
// ...
```

This implementation leverages a `Deferred` object to ensure `renderInit` waits until all components are registered through `ReactOnRails.register`. Could this approach address the issue effectively?

**Reply by AbanoubGhadban** [reply, marked as answer]
@kotarella1110 We already implemented an approach similar to your approach here https://github.com/shakacode/react_on_rails/pull/1656. The PR will be merged very soon. But for backward compatibility, you need to pass `force_load: true` on the server side for any component you want to hydrate immediately.

**Comment by justin808** [comment]
@kotarella1110, if you think my team and I can help you further with https://readyfor.jp, please feel free to [book a time](https://meetings.hubspot.com/justingordon/30-minute-consultation).

**Comment by justin808** [comment]
@kotarella1110 we made major updates for v15. Let us know how that plays into your discussion.
